### PR TITLE
Use atomic builtins

### PR DESCRIPTION
**Pull request by RajalakshmiS**
1.Replaces __sync_fetch_and_and, __sync_fetch_and_or with __atomic_fetch_and
and __atomic_fetch_or builtins.
2.Correct SPHSinglePCQueuePrefetch() function name.

Signed-off-by:  Rajalakshmi Srinivasaraghavan  <raji@linux.vnet.ibm.com>

	* src/bitv.c (bitv_alloc_marked): Use atomic builtins.
	* src/bitv.c (bitv_aligned_alloc_marked): Likewise.
	* src/bitv.c (bitv_dealloc_internal): Likewise.
	* src/bitv.c (GCC_VERSION): Define.
	* src/sasatom.h (fetch_and_and): New function.
	* src/sasatom.h (fetch_and_or): New function.
	* src/sasatom.h (fetch_and_and_long): New function.
	* src/sasatom.h (fetch_and_or_long): New function.
	* src/sphdirectpcqueue.h (SPHLFEntryDirectComplete): Comment typo
	correction.
	* src/sphlflogger.cpp (SPHSinglePCQueueSetCachePrefetch): Use atomic
	builtins.
	* src/sphlflogger.cpp (SPHSinglePCQueuePrefetch): Name correction.
	* src/sphsinglepcqueue.cpp (SPHLFLoggerAllocTimeStamped): Use atomic
	builtins.
	* src/sphsinglepcqueue.cpp (SPHLFLoggerAllocStrideTimeStamped):
	Likewise.
	* src/sphsinglepcqueue.cpp (SPHLFLoggerSetCachePrefetch): Likewise.

### CODE REVIEW

File: src/sasatom.h [hunk 1]
```diff
@@ -89,6 +91,77 @@ fetch_and_add(void *pointer, long int delta)
...
+ *
+ * Performs the atomic operation:
+ * { tmp = *pointer; *pointer = tmp & delta; return *pointer }
+ *
+ * Returns \a *pointer before update.
+ */
+static inline long int
+fetch_and_and(unsigned int *pointer, int delta)
+{
+#if GCC_VERSION >= 40700
```
Comment (tuliom) [inline src/sasatom.h]: What about testing for __ATOMIC_RELAXED instead of GCC_VERSION?

File: src/sphsinglepcqueue.cpp [hunk 2]
```diff
@@ -1232,14 +1232,14 @@ SPHSinglePCQueueSetCachePrefetch (SPHSinglePCQueue_t queue)
 	  touchptr += pagesz;
 	  touch += *touchptr;
 #if 0
```
Comment (tuliom) [inline src/sphsinglepcqueue.cpp]: I know you aren't touching this, but why #if 0?
Why not #ifdef PRINT_DEBUG or #ifdef __SASDebugPrint__?

File: src/sasatom.h [hunk 3]
```diff
@@ -89,6 +91,77 @@ fetch_and_add(void *pointer, long int delta)
...
+ * Performs the atomic operation:
+ * { tmp = *pointer; *pointer = tmp | delta; return *pointer }
+ *
+ * Returns \a *pointer before update.
+ */
+static inline long int
+fetch_and_or_long(unsigned long *pointer, long int delta)
+{
+#if GCC_VERSION >= 40700
+  return __atomic_fetch_or(pointer, delta, __ATOMIC_ACQUIRE);
```
Comment (tuliom) [inline src/sasatom.h]: Why are all of these __ATOMIC_ACQUIRE?
Aren't any of these __ATOMIC_ACQ_REL?

File: src/sasatom.h (hunk 1 above)
Comment (RajalakshmiS) [inline src/sasatom.h]: Checked in other libraries.Availability of built-in functions are handled using gcc version.

File: src/sphsinglepcqueue.cpp (hunk 2 above)
Comment (RajalakshmiS) [inline src/sphsinglepcqueue.cpp]: Yes. This has to be cleaned in all the files.Will do that in a separate patch.

File: src/sasatom.h (hunk 3 above)
Comment (RajalakshmiS) [inline src/sasatom.h]: Ack. Will change and send updated patch.

**Comment by tuliom** [comment]
LGTM

**Comment by RajalakshmiS** [comment]
Merged
